Keep a Changelog (#467)
open-telemetry/opentelemetry-dotnet-instrumentation: docs/README.md

- core components: [`1.2.0-rc4`](https://github.com/open-telemetry/opentelemetry-dotnet/releases/tag/core-1.2.0-rc4)
- non-core components: [`1.0.0-rc9.1`](https://github.com/open-telemetry/opentelemetry-dotnet/releases/tag/1.0.0-rc9.1)
- `System.Diagnostics.DiagnosticSource`: [`6.0.0`](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/6.0.0)

To auto-instrument applications, the OpenTelemetry .NET Auto-Instrumentation does the following:

1. Injects and configures the [OpenTelemetry .NET SDK](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/README.md#opentelemetry-net-sdk)
   into the application.
2. Adds [OpenTelemetry Instrumentation](https://opentelemetry.io/docs/concepts/instrumenting/)
   to key packages and APIs used by the application.

In addition, you can use the OpenTelemetry .NET Auto-Instrumentation as a .NET Profiler
to inject additional instrumentations at runtime, using a technique known as [monkey-patching](https://en.wikipedia.org/wiki/Monkey_patch).
This lets you instrument specific packages or APIs that does not provide the necessary hooks

> to generate .NET instrumentation packages using `System.Diagnostics.DiagnosticSource`.

Reviewer: Sorry that I am late to this change, but I think we should revisit this part of the README. I can imagine that if a developer downloads the package and reads this paragraph, they might get the idea that they can dynamically add instrumentation to their application, which is inaccurate. They cannot add new instrumentation dynamically, they can only use the instrumentation provided by the repo. Instead, I suggest rephrasing to say something like this:

> In addition, you can enable the OpenTelemetry .NET Auto-Instrumentation as a .NET Profiler
to inject additional instrumentations at runtime, using a technique known as [monkey-patching](https://en.wikipedia.org/wiki/Monkey_patch). This adds the capability to generate OTEL traces for libraries that do not generatie diagnostic information out of the box through standard providers like `System.Diagnostics.DiagnosticSource`, `System.Diagnostics.Activity` and the OpenTelemetry .NET SDK.

Reviewer 2: Should we also qualify that the additional instrumentation must be a part of this project? I can see the phrase `inject additional instrumentations at runtime` being misinterpreted as dynamic instrumentation.

Reviewer: Based off your suggestion Chris, here's another wording that we could use:

> In addition, you can enable the OpenTelemetry .NET Auto-Instrumentation as a .NET Profiler
to inject additional instrumentations at runtime, using a technique known as [monkey-patching](https://en.wikipedia.org/wiki/Monkey_patch). When enabled, the OpenTelemetry .NET Auto-Instrumentation will invoke instrumentations that are written in this repo to generate OpenTelemetry traces for libraries that do not already generate traces through the OpenTelemetry .NET SDK.

Does that capture your intent? The wording can still be improved, but maybe this is closer to what you were looking for?

Reviewer 2: That seems to capture the intent. We can always change it again later as we get more feedback.

Reviewer 3: @theletterf could please review the suggestion above before we incorporate it? Better to "review" it now and do a quick PR to improve just that part.

Reviewer 4: Looks great! Will edit in a follow-up PR.

Reviewer 4: See https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/483

This project adds [OpenTelemetry instrumentation](https://opentelemetry.io/docs/concepts/instrumenting/#automatic-instrumentation)
to .NET applications without having to modify their source code.

> It is built on top of [OpenTelemetry .NET](https://github.com/open-telemetry/opentelemetry-dotnet):

Reviewer 4: ```suggestion
The instrumentation is built on top of [OpenTelemetry .NET](https://github.com/open-telemetry/opentelemetry-dotnet):
```

Author: ```suggestion
OpenTelemetry .NET Auto-Instrumentation is built on top of [OpenTelemetry .NET](https://github.com/open-telemetry/opentelemetry-dotnet):
```

> - core components: [`1.2.0-rc4`](https://github.com/open-telemetry/opentelemetry-dotnet/releases/tag/core-1.2.0-rc4)
> - non-core components: [`1.0.0-rc9.1`](https://github.com/open-telemetry/opentelemetry-dotnet/releases/tag/1.0.0-rc9.1)

Reviewer 4: ```suggestion
- Core components: [`1.2.0-rc4`](https://github.com/open-telemetry/opentelemetry-dotnet/releases/tag/core-1.2.0-rc4)
- Non-core components: [`1.0.0-rc9.1`](https://github.com/open-telemetry/opentelemetry-dotnet/releases/tag/1.0.0-rc9.1)
```

Could there be a confusion around the word "core" here (as in .NET Core)? Alternative words:

- Main and secondary
- Primary and secondary

Author: This is a term used in OTel .NET. Currently using any different word can bring more confusion. See: https://github.com/open-telemetry/opentelemetry-dotnet/releases

FYI @cijothomas

- `System.Diagnostics.DiagnosticSource`: [`6.0.0`](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/6.0.0)

> To auto-instrument applications, OpenTelemetry .NET Auto-Instrumentation:

Reviewer 4: ```suggestion
To auto-instrument applications, the OpenTelemetry .NET Auto-Instrumentation does the following:
```

1. Injects and configures the [OpenTelemetry .NET SDK](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/README.md#opentelemetry-net-sdk)
   into the application.
2. Adds [OpenTelemetry Instrumentation](https://opentelemetry.io/docs/concepts/instrumenting/)
   to key packages and APIs used by the application.

Moreover, OpenTelemetry .NET Auto-Instrumentation can by used as .NET Profiler
in order to inject additional instrumentations at runtime
using a technique known as [monkey-patching](https://en.wikipedia.org/wiki/Monkey_patch).
This allows to instrument specific packages or APIs that does not provide the necessary hooks
to generate .NET instrumentation packages using `System.Diagnostics.DiagnosticSource`.

> See [examples](../examples) which demonstrates how different instrumentation scenarios
> are covered by OpenTelemetry .NET Auto-Instrumentation.

Reviewer 4: ```suggestion
See the [examples](../examples) for demonstrations of different instrumentation scenarios
covered by the OpenTelemetry .NET Auto-Instrumentation.
```
